src/Core/ValueMapperTrait.php

```diff
@@ -0,0 +1,62 @@
+<?php
+/**
+ * Copyright 2017 Google Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+namespace Google\Cloud\Core;
+
+/**
+ * Provides common value mapping methods.
+ */
+trait ValueMapperTrait
+{
+    /**
+     * Convert a timestamp (represented as a string or an array) to a Timestamp
+     * class with nanosecond support.
+     *
+     * @param string|array $timestamp The timestamp as a string or an array.
+     * @param string $returnType The type to create and return. Any object used
+     *        must provide a constructor compatible with {@see Google\Cloud\Core\Timestamp}.
+     *        **Defaults to** `Google\Cloud\Core\Timestamp`.
+     * @return mixed
+     */
+    public function createTimestampWithNanos($timestamp, $returnType = Timestamp::class)
+    {
+        $nanos = 0;
+        if (is_array($timestamp)) {
+            $timestamp = $timestamp + [
+                'seconds' => 0,
+                'nanos' => 0
+            ];
+
+            $dt = \DateTimeImmutable::createFromFormat('U', $timestamp['seconds']);
+            $nanos = $timestamp['nanos'];
+        } else {
+            $nanoRegex = '/(?:\.(\d{1,9})Z)|(?:Z)/';
+
+            $matches = [];
+            preg_match($nanoRegex, $timestamp, $matches);
+            $timestamp = preg_replace($nanoRegex, '.000000Z', $timestamp);
```

Comment:
PHP only has native support for precision down to microseconds. Due to this, we record the nanoseconds provided by a google cloud Timestamp outside of the PHP DateTime object (see ln 60 below). When the Timestamp is serialized for sending back to the API, the stored nanoseconds are reapplied. We do this to avoid any loss of precision or accidental data mutation in APIs such as Cloud Spanner.

Reply:
Thanks for the explanation, I didn't actually know that PHP has such limited precision. In the other clients, we actually do strip precision if the user fetches and resaves a timestamp, but at least those are stored with millisecond precision. Furthermore, this is only true for timestamps within Documents - system timestamps such as the updateTime on a document are treated differently (Node just stores them as ISO 8601 strings).

We do this to be able to support native time types in the various languages, while preserving the ability to op-lock on updateTime. If PHP isn't all centered around a a single date type (such as http://php.net/manual/en/class.datetime.php), then it might make sense to use your own type here (especially given the microsecond precision). I can't give you advice on what path to prefer here and I can see that there's rationale for sticking with Google Core's timestamp type.